File: packages/sdk/kotlin/client/src/commonMain/kotlin/dev/edgebase/sdk/client/PushClient.kt
// EdgeBase Kotlin SDK — Push notification client (KMP).
//
// Client-side push notification management.
// Delegates platform-specific operations to PlatformPush (expect/actual).
///§9/§10, #130: KMP 전환.

package dev.edgebase.sdk.client

import dev.edgebase.sdk.core.*

/**
 * Client-side push notification management.
 *
 * Usage:
 * ```kotlin
 * client.push.register()                              // auto platform token
 * client.push.register(mapOf("topic" to "news"))      // with metadata
 * ```
 */
class PushClient(
    private val client: HttpClient,
    private val platform: PlatformPush = PlatformPush()
) {
    private val messageListeners = mutableListOf<(Map<String, Any?>) -> Unit>()
    private val openedAppListeners = mutableListOf<(Map<String, Any?>) -> Unit>()

    private var cachedDeviceId: String? = null
    private var cachedToken: String? = null

    private fun getOrCreateDeviceId(): String {
        cachedDeviceId?.let { return it }
        val id = platformUuid()
        cachedDeviceId = id
        return id
    }

    /**
     * Register for push notifications.
     * Auto-acquires platform push token, caches, sends to server only on change (§9).
     */
    suspend fun register(metadata: Map<String, Any?>? = null) {
        // 1. Request permission
        val perm = requestPermission()
        if (perm != "granted") return

        // 2. Get FCM token
        val tokenPair = platform.getToken(client) ?: return
        val token = tokenPair.first

        // 3. Check cache — skip if unchanged (§9), unless metadata provided
        if (cachedToken == token && metadata == null) return

        // 4. Register with server — auto-collect deviceInfo
        val deviceId = getOrCreateDeviceId()
        val deviceInfo = platform.getDeviceInfo()
        val body = mutableMapOf<String, Any?>(
            "deviceId" to deviceId,
            "token" to token,
            "platform" to platform.getPlatformName(),
            "deviceInfo" to deviceInfo
        )
        if (metadata != null) body["metadata"] = metadata
        client.post("/push/register", body)
        cachedToken = token
    }

    /**
     * Unregister current device (or a specific device by ID).
     */
    suspend fun unregister(deviceId: String? = null) {
        val id = deviceId ?: getOrCreateDeviceId()
        client.post("/push/unregister", mapOf("deviceId" to id))
        cachedToken = null
    }

    /** Listen for push messages in foreground. */
    fun onMessage(callback: (Map<String, Any?>) -> Unit) {
        messageListeners.add(callback)
    }

    /** Listen for notification taps that opened the app. */
    fun onMessageOpenedApp(callback: (Map<String, Any?>) -> Unit) {
        openedAppListeners.add(callback)
    }

    /** Get notification permission status. Returns "granted"|"denied"|"notDetermined". */
    fun getPermissionStatus(): String {
        return platform.getPermissionStatus()
    }

    /** Request notification permission from the user. */
    suspend fun requestPermission(): String {
        return platform.requestPermission()
    }

    /**
     * Subscribe to a push notification topic.
     * Mobile: Firebase SDK directly. Web: server-side via IID API.
     */
    suspend fun subscribeTopic(topic: String) {
        platform.subscribeTopic(topic, client)
    }

    /**
     * Unsubscribe from a push notification topic.
     */
    suspend fun unsubscribeTopic(topic: String) {
        platform.unsubscribeTopic(topic, client)
    }

    /** Override token acquisition for headless or custom-platform integrations. */
    fun setTokenProvider(provider: suspend () -> String) {
        platform.setTokenProvider(provider)
    }

    /**
     * Override permission status/request flows for headless or custom-platform integrations.
     * Either closure may be omitted to keep the platform default for that branch.
     */
    fun setPermissionProvider(
        getPermissionStatus: (() -> String)? = null,
        requestPermission: (suspend () -> String)? = null
    ) {
        platform.setPermissionStatusProvider(getPermissionStatus)
        platform.setPermissionRequester(requestPermission)
    }

    /** Override topic subscription plumbing for headless or custom-platform integrations. */
    fun setTopicProvider(
        subscribe: suspend (String) -> Unit,
        unsubscribe: suspend (String) -> Unit
    ) {
        platform.setTopicProvider(subscribe, unsubscribe)
    }

    /** Dispatch a foreground message to registered listeners. Called from platform push service. */
    fun dispatchMessage(message: Map<String, Any?>) {
        for (cb in messageListeners) cb(message)
    }

    /** Dispatch a notification-opened event. Called from platform app lifecycle. */
    fun dispatchMessageOpenedApp(message: Map<String, Any?>) {
        for (cb in openedAppListeners) cb(message)
    }
}

File: packages/sdk/dart/packages/flutter/lib/src/push_client.dart
/// PushClient — Push notification management for Flutter.
///
/// Auto-acquires push token via FirebaseMessaging (Android: FCM, iOS: APNs).
/// Developer only needs google-services.json (Android) / GoogleService-Info.plist (iOS).
///
/// ```dart
/// await client.push.register();                       // auto token
/// await client.push.register(metadata: {'topic': 'news'}); // with metadata
/// ```
import 'dart:async';
import 'dart:io' show Platform;
import 'package:firebase_messaging/firebase_messaging.dart';
import 'package:edgebase_core/src/http_client.dart';
import 'package:edgebase_core/src/generated/api_core.dart';

/// Push notification platform type.
enum PushPlatform { ios, android, web, macos, windows }

/// Client-side push notification management.
/// Auto-acquires token via FirebaseMessaging — no tokenProvider needed.
class PushClient {
  final GeneratedDbApi _core;
  final List<void Function(Map<String, dynamic>)> _messageListeners = [];
  final List<void Function(Map<String, dynamic>)> _openedAppListeners = [];
  Future<String> Function()? _tokenProvider;
  FutureOr<String> Function()? _permissionStatusProvider;
  Future<String> Function()? _permissionRequester;
  Future<void> Function(String topic)? _topicSubscriber;
  Future<void> Function(String topic)? _topicUnsubscriber;
  PushPlatform? _platformOverride;

  // Internal cache
  String? _cachedDeviceId;
  String? _cachedToken;

  PushClient(HttpClient http) : _core = GeneratedDbApi(http) {
    // Auto-listen for foreground messages
    FirebaseMessaging.onMessage.listen((RemoteMessage message) {
      final msg = <String, dynamic>{
        'title': message.notification?.title,
        'body': message.notification?.body,
        'data': message.data,
      };
      for (final cb in _messageListeners) { cb(msg); }
    });

    // Auto-listen for notification taps
    FirebaseMessaging.onMessageOpenedApp.listen((RemoteMessage message) {
      final msg = <String, dynamic>{
        'title': message.notification?.title,
        'body': message.notification?.body,
        'data': message.data,
      };
      for (final cb in _openedAppListeners) { cb(msg); }
    });
  }

  /// Override token acquisition for headless or custom-platform integrations.
  void setTokenProvider(
    Future<String> Function() provider, {
    PushPlatform platform = PushPlatform.web,
  }) {
    _tokenProvider = provider;
    _platformOverride = platform;
  }

  /// Override permission status/request flows for headless or custom-platform integrations.
  void setPermissionProvider({
    FutureOr<String> Function()? getPermissionStatus,
    Future<String> Function()? requestPermission,
  }) {
    _permissionStatusProvider = getPermissionStatus;
    _permissionRequester = requestPermission;
  }

  /// Override topic subscription plumbing for headless or custom-platform integrations.
  void setTopicProvider({
    required Future<void> Function(String topic) subscribe,
    required Future<void> Function(String topic) unsubscribe,
  }) {
    _topicSubscriber = subscribe;
    _topicUnsubscriber = unsubscribe;
  }

  String _getOrCreateDeviceId() {
    _cachedDeviceId ??= _generateUuid();
    return _cachedDeviceId!;
  }

  String _generateUuid() {
    final r = DateTime.now().microsecondsSinceEpoch;
    return '${r.toRadixString(16).padLeft(12, '0')}-${r.hashCode.toRadixString(16).padLeft(8, '0')}';
  }

  PushPlatform _detectPlatform() {
    if (_platformOverride != null) return _platformOverride!;
    try {
      if (Platform.isIOS) return PushPlatform.ios;
      if (Platform.isAndroid) return PushPlatform.android;
      if (Platform.isMacOS) return PushPlatform.macos;
      if (Platform.isWindows) return PushPlatform.windows;
    } catch (_) {
      // Platform not available (web)
    }
    return PushPlatform.web;
  }

  /// Register for push notifications.
  /// Auto-acquires token via FirebaseMessaging (FCM on Android, APNs on iOS).
  /// Caches token, sends to server only on change (§9).
  Future<void> register({Map<String, dynamic>? metadata}) async {
    // 1. Request permission
    final perm = await requestPermission();
    if (perm != 'granted') return;

    // 2. Get token from FirebaseMessaging — auto handles Android (FCM) + iOS (APNs)
    final token = _tokenProvider != null
        ? await _tokenProvider!()
        : await FirebaseMessaging.instance.getToken();
    if (token == null || token.isEmpty) {
      throw StateError('Failed to get push token. Ensure Firebase is configured.');
    }

    // 3. Check cache — skip if unchanged (§9), unless metadata provided
    if (_cachedToken == token && metadata == null) return;

    // 4. Register with server — auto-collect deviceInfo
    final deviceId = _getOrCreateDeviceId();
    final deviceInfo = _collectDeviceInfo();
    final body = <String, dynamic>{
      'deviceId': deviceId,
      'token': token,
      'platform': _detectPlatform().name,
      'deviceInfo': deviceInfo,
    };
    if (metadata != null) body['metadata'] = metadata;
    await _core.pushRegister(body);
    _cachedToken = token;
  }

  Map<String, String> _collectDeviceInfo() {
    final info = <String, String>{};
    try {
      info['osVersion'] = '${Platform.operatingSystem} ${Platform.operatingSystemVersion}';
      info['locale'] = Platform.localeName;
    } catch (_) {
      info['osVersion'] = 'web';
    }
    return info;
  }

  /// Unregister current device (or a specific device by ID).
  Future<void> unregister([String? deviceId]) async {
    final id = deviceId ?? _getOrCreateDeviceId();
    await _core.pushUnregister({'deviceId': id});
    _cachedToken = null;
  }

  /// Listen for push messages in foreground.
  void onMessage(void Function(Map<String, dynamic> message) callback) {
    _messageListeners.add(callback);
  }

  /// Listen for notification taps that opened the app.
  void onMessageOpenedApp(void Function(Map<String, dynamic> message) callback) {
    _openedAppListeners.add(callback);
  }

  /// Get notification permission status.
  Future<String> getPermissionStatus() async {
    if (_permissionStatusProvider != null) {
      return await Future<String>.value(_permissionStatusProvider!.call());
    }
    final settings = await FirebaseMessaging.instance.getNotificationSettings();
    switch (settings.authorizationStatus) {
      case AuthorizationStatus.authorized:
      case AuthorizationStatus.provisional:
        return 'granted';
      case AuthorizationStatus.denied:
        return 'denied';
      default:
        return 'notDetermined';
    }
  }

  /// Subscribe to an FCM topic.
  /// Uses Firebase SDK directly — mobile only.
  Future<void> subscribeTopic(String topic) async {
    if (_topicSubscriber != null) {
      await _topicSubscriber!(topic);
      return;
    }
    await FirebaseMessaging.instance.subscribeToTopic(topic);
  }

  /// Unsubscribe from an FCM topic.
  Future<void> unsubscribeTopic(String topic) async {
    if (_topicUnsubscriber != null) {
      await _topicUnsubscriber!(topic);
      return;
    }
    await FirebaseMessaging.instance.unsubscribeFromTopic(topic);
  }

  /// Request notification permission from the user.
  Future<String> requestPermission() async {
    if (_permissionRequester != null) {
      return await _permissionRequester!();
    }
    final settings = await FirebaseMessaging.instance.requestPermission(
      alert: true,
      badge: true,
      sound: true,
    );
    return settings.authorizationStatus == AuthorizationStatus.authorized ||
           settings.authorizationStatus == AuthorizationStatus.provisional
        ? 'granted'
        : 'denied';
  }

  /// Dispatch a foreground message to registered listeners.
  void dispatchMessage(Map<String, dynamic> message) {
    for (final cb in _messageListeners) {
      cb(message);
    }
  }

  /// Dispatch a notification-opened event to registered listeners.
  void dispatchMessageOpenedApp(Map<String, dynamic> message) {
    for (final cb in _openedAppListeners) {
      cb(message);
    }
  }
}
